File: packages/core/src/dto/species.ts
import { z } from "zod";
import { GROWTH_RATES, TOLERANCE_LEVELS } from "../constants.js";

export const NpkSchema = z.object({
  n: z.number().min(0),
  p: z.number().min(0),
  k: z.number().min(0),
});
export type Npk = z.infer<typeof NpkSchema>;

export const SpeciesIdealSchema = z.object({
  luxMin: z.number().min(0),
  luxIdeal: z.number().min(0),
  luxMax: z.number().min(0),
  tempMinC: z.number(),
  tempMaxC: z.number(),
  humidityMin: z.number().min(0).max(100),
  humidityMax: z.number().min(0).max(100),
  phMin: z.number(),
  phMax: z.number(),
  npk: NpkSchema,
  waterIntervalDays: z.number().positive(),
  waterAmountMl: z.number().positive(),
  fertIntervalDays: z.number().positive(),
  mistIntervalDays: z.number().nonnegative(),
  repotIntervalMonths: z.number().positive(),
  rotateIntervalDays: z.number().positive(),
});
export type SpeciesIdeal = z.infer<typeof SpeciesIdealSchema>;

export const SpeciesToleranceSchema = z.object({
  drought: z.enum(TOLERANCE_LEVELS),
  shade: z.enum(TOLERANCE_LEVELS),
  cold: z.enum(TOLERANCE_LEVELS),
});
export type SpeciesTolerance = z.infer<typeof SpeciesToleranceSchema>;

export const SpeciesGrowthSchema = z.object({
  rate: z.enum(GROWTH_RATES),
  maxHeightCm: z.number().positive(),
});
export type SpeciesGrowth = z.infer<typeof SpeciesGrowthSchema>;

export const SpeciesSchema = z.object({
  slug: z.string().regex(/^[a-z0-9]+(-[a-z0-9]+)*$/),
  commonNames: z.array(z.string()).min(1),
  scientificName: z.string(),
  family: z.string(),
  toxicity: z.object({
    pets: z.boolean(),
    note: z.string().optional(),
  }),
  ideal: SpeciesIdealSchema,
  tolerance: SpeciesToleranceSchema,
  growth: SpeciesGrowthSchema,
});
export type Species = z.infer<typeof SpeciesSchema>;

File: apps/android/app/src/main/java/org/silvae/domain/care/Species.kt
package org.silvae.domain.care

import kotlinx.serialization.Serializable

/** Mirrors packages/core/src/dto/species.ts — parsed from the bundled data/species JSON assets. */
@Serializable
data class Npk(val n: Double, val p: Double, val k: Double)

@Serializable
data class SpeciesIdeal(
    val luxMin: Double,
    val luxIdeal: Double,
    val luxMax: Double,
    val tempMinC: Double,
    val tempMaxC: Double,
    val humidityMin: Double,
    val humidityMax: Double,
    val phMin: Double,
    val phMax: Double,
    val npk: Npk,
    val waterIntervalDays: Double,
    val waterAmountMl: Double,
    val fertIntervalDays: Double,
    val mistIntervalDays: Double,
    val repotIntervalMonths: Double,
    val rotateIntervalDays: Double,
)

@Serializable
data class SpeciesTolerance(val drought: String, val shade: String, val cold: String)

@Serializable
data class SpeciesGrowth(val rate: String, val maxHeightCm: Double)

@Serializable
data class SpeciesToxicity(val pets: Boolean, val note: String? = null)

@Serializable
data class Species(
    val slug: String,
    val commonNames: List<String>,
    val scientificName: String,
    val family: String,
    val toxicity: SpeciesToxicity,
    val ideal: SpeciesIdeal,
    val tolerance: SpeciesTolerance,
    val growth: SpeciesGrowth,
)
